### PR TITLE
Add new IPC for trace dma free

### PR DESCRIPTION
**Pull request by ranj063**
FW PR https://github.com/thesofproject/sof/pull/4849

### CODE REVIEW

File: sound/soc/sof/trace.c [hunk 1]
```diff
@@ -549,6 +553,20 @@ void snd_sof_release_trace(struct snd_sof_dev *sdev)
 		dev_err(sdev->dev,
 			"error: snd_sof_dma_trace_trigger: stop: %d\n", ret);
 
+	/*
+	 * stop and free trace DMA in the DSP. TRACE_DMA_FREE is only supported from
+	 * ABI 3.20.0 onwards
+	 */
+	if (v->abi_version >= SOF_ABI_VER(3, 20, 0)) {
+		params.hdr.size = sizeof(struct sof_ipc_dma_trace_params);
+		params.hdr.cmd |= SOF_IPC_TRACE_DMA_FREE;
```
Comment (ujfalusi) [inline sound/soc/sof/trace.c]: `params.hdr.cmd = SOF_IPC_GLB_TRACE_MSG | SOF_IPC_TRACE_DMA_FREE;`

File: sound/soc/sof/trace.c [hunk 2]
```diff
@@ -539,6 +539,10 @@ EXPORT_SYMBOL(snd_sof_trace_notify_for_error);
 
 void snd_sof_release_trace(struct snd_sof_dev *sdev)
 {
+	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
+	struct sof_ipc_fw_version *v = &ready->version;
+	struct sof_ipc_dma_trace_params_ext params;
```
Comment (ujfalusi) [inline sound/soc/sof/trace.c]: It should be enough to send simple message, no?
`struct sof_ipc_cmd_hdr msg;`

**Review by ujfalusi — CHANGES_REQUESTED** [review]

**Comment by ujfalusi** [comment]
@ranj063, while at it, can you also update the `ipc_log_header()` to handle this new message?

File: sound/soc/sof/trace.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: a simple message would need to be parsed separately. if it is a trace ipc message, the glb type would already be handled in the FW

File: sound/soc/sof/trace.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: done

**Comment by ranj063** [comment]
> @ranj063, while at it, can you also update the `ipc_log_header()` to handle this new message?

done. thanks @ujfalusi

File: sound/soc/sof/trace.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: sorry @ujfalusi i misunderstood the comment. you are right. modified now

File: sound/soc/sof/trace.c [hunk 3]
```diff
@@ -539,6 +539,10 @@ EXPORT_SYMBOL(snd_sof_trace_notify_for_error);
 
 void snd_sof_release_trace(struct snd_sof_dev *sdev)
 {
+	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
+	struct sof_ipc_fw_version *v = &ready->version;
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply ipc_reply;
```
Comment (ujfalusi) [inline sound/soc/sof/trace.c:545]: nitpick: can you swap these?

or move them local under the `if(ABI)` ?

**Review by ujfalusi — APPROVED** [review]
@ranj063, looks good, thanks. Only one nitpick if you want to fix it.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:545]: @ranj063 let's do this change suggested by @ujfalusi  with a fixup

File: sound/soc/sof/trace.c [hunk 4]
```diff
@@ -549,6 +553,20 @@ void snd_sof_release_trace(struct snd_sof_dev *sdev)
...
+	 * ABI 3.20.0 onwards
+	 */
+	if (v->abi_version >= SOF_ABI_VER(3, 20, 0)) {
+		hdr.size = sizeof(hdr);
+		hdr.cmd = SOF_IPC_GLB_TRACE_MSG | SOF_IPC_TRACE_DMA_FREE;
+
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &ipc_reply, sizeof(ipc_reply));
+		if (ret < 0)
+			dev_err(sdev->dev, "DMA_TRACE_FREE failed with error: %d\n", ret);
```
Comment (plbossart) [inline sound/soc/sof/trace.c:567]: return ret?

or add a comment that on why we don't return?

This can be done with a fixup PR.

Comment (ujfalusi) [reply sound/soc/sof/trace.c:567]: I think the intention is to continue to the host side release, even if the IPC fails, in this case this should be `dev_warn()`?

**Pull request merged**